File: app/src/main/java/berlindroid/zethree/util/LowQualityUpdateDownloader.java
package berlindroid.zethree.util;

import android.app.Activity;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class LowQualityUpdateDownloader {

    private final Activity activity;

    public LowQualityUpdateDownloader(Activity activity) {
        super();
        this.activity = activity;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class Asset {
        @JsonProperty("browser_download_url") public String url;
        @JsonProperty("content_type") public String contentType;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class Release {
        @JsonProperty("published_at") public Date publishedAt;
        @JsonProperty("assets") public List<Asset> assets;
        @JsonProperty("tag_name") public String tag;
    }

    public static class ResolvedApkUrl {
        public String url;
        public String appVersion;

        private ResolvedApkUrl(String url, String appVersion) {
            super();
            this.url = url;
            this.appVersion = appVersion;
        }
    }

    @SuppressWarnings({"SameParameterValue", "OptionalGetWithoutIsPresent"})
    public ResolvedApkUrl getLatestApkUrl(String url) {
        try {
            String content = getJSON(url, 5000);
            if (content == null) return null;
            ObjectMapper mapper = new ObjectMapper();
            List<Release> releases = mapper.readValue(
                content,
                new TypeReference<List<Release>>() { }
            );

            String apkType = "application/vnd.android.package-archive";
            Predicate<Asset> isApk = asset -> asset.contentType.equals((apkType));
            List<ResolvedApkUrl> allAssetData = releases.stream()
                .filter(release -> release.assets.stream().anyMatch(isApk))
                .sorted(Comparator.comparing(o -> o.publishedAt))
                .map(release -> {
                    String apkUrl = release.assets.stream().filter(isApk).findFirst().get().url;
                    String appVersion = release.tag.replaceFirst("v", "");
                    return new ResolvedApkUrl(apkUrl, appVersion);
                })
                .collect(Collectors.toList());

            // sorted, just take the last (latest)
            return allAssetData.get(allAssetData.size() - 1);
        } catch (Throwable t) {
            failAndClose(t);
            return null;
        }
    }

    @SuppressWarnings("SameParameterValue")
    private String getJSON(String url, int timeout) {
        HttpURLConnection c = null;
        try {
            URL u = new URL(url);
            c = (HttpURLConnection) u.openConnection();
            c.setRequestMethod("GET");
            c.setUseCaches(false);
            c.setAllowUserInteraction(false);
            c.setConnectTimeout(timeout);
            c.setReadTimeout(timeout);
            c.setRequestProperty("Accept", "application/vnd.github.v3+json");

//            c.setRequestProperty(
//                "Authorization",
//                "Bearer <YOUR_TOKEN_HERE>"
//            );
            c.connect();

            int status = c.getResponseCode();
            switch (status) {
                case 200:
                case 201:
                    InputStreamReader sr = new InputStreamReader(c.getInputStream());
                    BufferedReader br = new BufferedReader(sr);
                    StringBuilder sb = new StringBuilder();
                    String line;
                    while ((line = br.readLine()) != null) {
                        sb.append(line);
                        sb.append("\n");
                    }
                    br.close();
                    return sb.toString();
                default:
                    failAndClose(
                        new IllegalStateException("HTTP/" + status + ": " + c.getResponseMessage())
                    );
                    return null;
            }
        } catch (Throwable t) {
            failAndClose(t);
        } finally {
            if (c != null) {
                try {
                    c.disconnect();
                } catch (Throwable e) {
                    e.printStackTrace();
                }
            }
        }
        return null;
    }

    private void failAndClose(Throwable t) {
        new Handler(Looper.getMainLooper()).post(
            () -> {
                Toast.makeText(
                    activity,
                    "Failed, pfff...\n" + t.getMessage(),
                    Toast.LENGTH_LONG
                ).show();
                activity.finish();
            }
        );
    }

}

File: app/src/main/java/berlindroid/zethree/util/LowQualityUpdateDownloader.kt
package berlindroid.zethree.util

import android.util.Log
import android.util.Pair
import com.fasterxml.jackson.databind.ObjectMapper
import berlindroid.zethree.BuildConfig
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.core.type.TypeReference
import java.io.BufferedReader
import java.io.InputStreamReader
import java.lang.IllegalStateException
import java.lang.StringBuilder
import java.net.HttpURLConnection
import java.net.URL
import java.util.*
import java.util.function.Predicate
import java.util.stream.Collectors

object LowQualityUpdateDownloader {
    private const val RELEASES_URL =
        "https://api.github.com/repos/gdg-berlin-android/ZeThree/releases"
    private const val TIMEOUT = 5000

    fun resolveLatestApk(): ApkResolutionFactory {
        return try {
            val content = getJSON(RELEASES_URL, TIMEOUT)
                ?: return ApkResolutionFactory.withError("Network error")
            val releases: List<Release> = ObjectMapper().readValue(
                content, object : TypeReference<List<Release>>() {}
            )
            val apkType = "application/vnd.android.package-archive"
            val isApk = Predicate { asset: Asset -> asset.contentType == apkType }
            val assetDataSorted = releases.stream()
                .filter { release: Release -> release.assets.stream().anyMatch(isApk) }
                .sorted(Comparator.comparing { o: Release -> o.publishedAt })
                .map { release: Release ->
                    val apkUrl = release.assets.stream().filter(isApk).findFirst().get().url
                    val appVersion = release.tag.replaceFirst("v".toRegex(), "")
                    Pair(apkUrl, appVersion)
                }
                .collect(Collectors.toList())

            // sorted, just take the last (latest)
            val lastReleaseData = assetDataSorted[assetDataSorted.size - 1]
            val apkUrl = lastReleaseData.first
            val appVersion = lastReleaseData.second
            if (BuildConfig.VERSION_NAME == appVersion) {
                ApkResolutionFactory.withVersion(appVersion)
            } else {
                ApkResolutionFactory.withUrl(apkUrl)
            }
        } catch (t: Throwable) {
            Log.e(LowQualityUpdateDownloader::class.java.simpleName, "Failed", t)
            ApkResolutionFactory.withError(t.message)
        }
    }

    private fun getJSON(url: String, timeout: Int): String? {
        var c: HttpURLConnection? = null
        return try {
            val u = URL(url)
            c = (u.openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                useCaches = false
                allowUserInteraction = false
                connectTimeout = timeout
                readTimeout = timeout
                setRequestProperty("Accept", "application/vnd.github.v3+json")
            }

//            c.setRequestProperty(
//                "Authorization",
//                "Bearer <YOUR_TOKEN_HERE>"
//            );
            c.connect()
            val successCodes = arrayOf(200, 201)
            when (val status = c.responseCode) {
                in successCodes -> {
                    val sr = InputStreamReader(c.inputStream)
                    val br = BufferedReader(sr)
                    val sb = StringBuilder()
                    var line: String?
                    while (br.readLine().also { line = it } != null) {
                        sb.append(line)
                        sb.append("\n")
                    }
                    br.close()
                    sb.toString()
                }
                else -> {
                    val t: Throwable =
                        IllegalStateException("HTTP/" + status + ": " + c.responseMessage)
                    Log.e(LowQualityUpdateDownloader::class.java.simpleName, "Failed", t)
                    null
                }
            }
        } catch (t: Throwable) {
            Log.e(LowQualityUpdateDownloader::class.java.simpleName, "Failed", t)
            null
        } finally {
            if (c != null) {
                try {
                    c.disconnect()
                } catch (e: Throwable) {
                    e.printStackTrace()
                }
            }
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private data class Asset(
        @JsonProperty("browser_download_url")
        val url: String,

        @JsonProperty("content_type")
        val contentType: String
    )

    @JsonIgnoreProperties(ignoreUnknown = true)
    private data class Release(
        @JsonProperty("published_at")
        val publishedAt: Date,

        @JsonProperty("assets")
        val assets: List<Asset>,

        @JsonProperty("tag_name")
        val tag: String
    )

    class ApkResolutionFactory private constructor() {
        var status = Status.FAILED
        var url: String? = null
        var appVersion: String? = null
        var errorMessage: String? = null

        companion object {
            fun withUrl(url: String?): ApkResolutionFactory {
                val r = ApkResolutionFactory()
                r.status = Status.APK_AVAILABLE
                r.url = url
                return r
            }

            fun withVersion(appVersion: String?): ApkResolutionFactory {
                val r = ApkResolutionFactory()
                r.status = Status.ALREADY_LATEST
                r.appVersion = appVersion
                return r
            }

            fun withError(error: String?): ApkResolutionFactory {
                val r = ApkResolutionFactory()
                r.status = Status.FAILED
                r.errorMessage = error
                return r
            }
        }
    }

    enum class Status {
        FAILED, APK_AVAILABLE, ALREADY_LATEST
    }
}
